optuna/samplers/cmaes.py

```diff
@@ -0,0 +1,369 @@
+            seed: Optional[int] = None,
+    ) -> None:
+
+        self._x0 = x0
+        self._sigma0 = sigma0
+        self._independent_sampler = (
+            independent_sampler or optuna.samplers.RandomSampler(seed=seed)
+        )
+        self._n_startup_trials = n_startup_trials
+        self._warn_independent_sampling = warn_independent_sampling
+        self._logger = optuna.logging.get_logger(__name__)
+        self._cma_rng = np.random.RandomState(seed)
+
+    def infer_relative_search_space(
+            self, study: 'optuna.Study', trial: 'optuna.structs.FrozenTrial',
+    ) -> Dict[str, BaseDistribution]:
+
+        # Import here to avoid circular imports without breaking a backward compatibility.
+        from optuna.samplers import intersection_search_space
+
+        search_space = {}  # type: Dict[str, BaseDistribution]
+        for name, distribution in intersection_search_space(study).items():
+            if distribution.single():
+                # `cma` cannot handle distributions that contain just a single value, so we skip
+                # them. Note that the parameter values for such distributions are sampled in
+                # `Trial`.
+                continue
+
+            if not isinstance(
+                    distribution,
+                    (
+                        optuna.distributions.UniformDistribution,
+                        optuna.distributions.LogUniformDistribution,
+                        optuna.distributions.DiscreteUniformDistribution,
+                        optuna.distributions.IntUniformDistribution,
+                    ),
+            ):
+                # Categorical distribution is unsupported.
+                continue
+            search_space[name] = distribution
+
+        return search_space
+
+    def sample_relative(
+            self,
+            study: 'optuna.Study',
+            trial: 'optuna.structs.FrozenTrial',
+            search_space: Dict[str, BaseDistribution],
+    ) -> Dict[str, Any]:
+
+        if len(search_space) == 0:
+            return {}
+
+        completed_trials = [
+            t
+            for t in study.get_trials(deepcopy=False)
+            if t.state == TrialState.COMPLETE
+        ]
+        if len(completed_trials) < self._n_startup_trials:
+            return {}
+
+        if len(search_space) == 1:
+            self._logger.info(
+                "`CmaEsSampler` only supports two or more dimensional continuous "
+                "search space. `{}` is used instead of `CmaEsSampler`.".format(
+                    self._independent_sampler.__class__.__name__
+                )
+            )
+            self._warn_independent_sampling = False
+            return {}
+
+        ordered_keys = [key for key in search_space]
+        ordered_keys.sort()
+
+        optimizer = self._restore_or_init_optimizer(
+            completed_trials, search_space, ordered_keys
+        )
+
+        if optimizer.dim != len(ordered_keys):
```

Comment:
`dim` property method is added from the latest release of `cmaes` package.
https://pypi.org/project/cmaes/0.3.2/